File: scripts/sync_local_business_jsonld.py
#!/usr/bin/env python3
"""Replace nutri-thrive-local-business-jsonld blocks sitewide with batch_local_seo.LOCAL_JSON_LD."""
from __future__ import annotations

import re
import sys
from pathlib import Path

_SCRIPTS = Path(__file__).resolve().parent
if str(_SCRIPTS) not in sys.path:
    sys.path.insert(0, str(_SCRIPTS))

import batch_local_seo as _bls

LOCAL_JSON_LD = _bls.LOCAL_JSON_LD
MARKER = _bls.MARKER
should_process = _bls.should_process

ROOT = Path(__file__).resolve().parents[1]

BLOCK_RE = re.compile(
    r"(?:<!--\s*Nutri Thrive LocalBusiness[^>]*-->\s*)?"
    r'<script type="application/ld\+json"\s+id="' + re.escape(MARKER) + r'"[\s\S]*?</script>',
    re.IGNORECASE,
)


def main() -> None:
    changed = []
    for path in sorted(ROOT.rglob("*.html")):
        if not should_process(path):
            continue
        raw = path.read_text(encoding="utf-8", errors="replace")
        if MARKER not in raw:
            continue
        new = BLOCK_RE.sub(LOCAL_JSON_LD, raw, count=1)
        if new != raw:
            path.write_text(new, encoding="utf-8")
            changed.append(path.relative_to(ROOT))
    print(f"Updated {len(changed)} files with LocalBusiness JSON-LD")
    for p in changed[:30]:
        print(" ", p)
    if len(changed) > 30:
        print(f" ... and {len(changed) - 30} more")


if __name__ == "__main__":
    main()

File: scripts/sync-local-business-jsonld.js
#!/usr/bin/env node
/**
 * Single source of truth for standalone LocalBusiness JSON-LD (id=nutri-thrive-local-business-jsonld).
 * Edit: shared/schema/nutrithrive-local-business.json
 * Run:  node scripts/sync-local-business-jsonld.js
 *
 * Replaces the comment + script block in every HTML file that already contains that id
 * (inline JSON-LD so crawlers do not depend on external script src).
 */

const fs = require("fs");
const path = require("path");

const REPO = path.join(__dirname, "..");
const SCHEMA = path.join(
  REPO,
  "shared/schema/nutrithrive-local-business.json",
);

const BLOCK_RE =
  /(?:<!--\s*Nutri Thrive LocalBusiness[\s\S]*?-->\s*)?<script type="application\/ld\+json" id="nutri-thrive-local-business-jsonld">[\s\S]*?<\/script>/g;

const IGNORE_DIRS = new Set([
  "node_modules",
  ".git",
  "dist",
  "build",
  ".next",
]);

function buildReplacement() {
  const raw = fs.readFileSync(SCHEMA, "utf8");
  const obj = JSON.parse(raw);
  const pretty = JSON.stringify(obj, null, 2);
  return (
    `<!-- Nutri Thrive LocalBusiness — source: shared/schema/nutrithrive-local-business.json (sync: node scripts/sync-local-business-jsonld.js) -->\n` +
    `<script type="application/ld+json" id="nutri-thrive-local-business-jsonld">\n${pretty}\n</script>`
  );
}

function walkHtml(dir, out) {
  let entries;
  try {
    entries = fs.readdirSync(dir, { withFileTypes: true });
  } catch {
    return;
  }
  for (const e of entries) {
    const full = path.join(dir, e.name);
    if (e.isDirectory()) {
      if (IGNORE_DIRS.has(e.name)) continue;
      walkHtml(full, out);
    } else if (e.isFile() && e.name.endsWith(".html")) {
      out.push(full);
    }
  }
}

function main() {
  const replacement = buildReplacement();
  const files = [];
  walkHtml(REPO, files);

  let changed = 0;
  let touched = 0;
  for (const abs of files) {
    let s = fs.readFileSync(abs, "utf8");
    if (!s.includes('id="nutri-thrive-local-business-jsonld"')) continue;
    const next = s.replace(BLOCK_RE, replacement);
    if (next === s) continue;
    fs.writeFileSync(abs, next, "utf8");
    changed++;
    touched++;
  }

  console.log(
    `Synced LocalBusiness JSON-LD from ${path.relative(REPO, SCHEMA)} into ${changed} file(s).`,
  );
  if (touched === 0) {
    console.log(
      "No files needed updates (all blocks already match, or id not found).",
    );
  }
}

main();
